File: Xamarin/Xamarin.Native/Xamarin.Native.Droid/Activities/PerformanceTestActivity.cs
using Android.App;
using Android.Content;
using Android.OS;
using Android.Runtime;
using Android.Views;
using Android.Widget;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Xamarin.Services;

namespace Xamarin.Native.Droid.Activities
{
    [Activity(Label = "PerformanceActivity")]
    public class PerformanceTestActivity : Activity
    {
        Stopwatch stopwatch;
        Button startButton;
        EditText digitsEntry;
        TextView ResultView;
        TextView TimeLabel;
        ProgressDialog progressDialog;

        protected override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);
            SetContentView(Resource.Layout.PerformanceTest);
            Title = GetString(Resource.String.menuPerformance);

            PerformanceTestService.Instance.PiCalculationCompleted += Instance_CalculationCompleted;

            digitsEntry = FindViewById<EditText>(Resource.Id.digitsEntry);
            ResultView = FindViewById<TextView>(Resource.Id.resultView);
            TimeLabel = FindViewById<TextView>(Resource.Id.timeLabel);
            startButton = FindViewById<Button>(Resource.Id.startButton);

            startButton.Click += StartCalculation;
        }

        private void StartCalculation(object sender, EventArgs e)
        {
            stopwatch = new Stopwatch();
            progressDialog = ProgressDialog.Show(this, "Przetwarzanie...", "");

            var digits = Convert.ToInt32(digitsEntry.Text);
            Task.Run(() =>
            {
                stopwatch.Start();
                PerformanceTestService.Instance.CalculatePi(digits);
            });
        }

        private void Instance_CalculationCompleted(string result)
        {
            stopwatch.Stop();
            RunOnUiThread(() =>
            {
                ResultView.Text = result;
                TimeLabel.Text = stopwatch.GetDurationInSeconds();
                progressDialog.Dismiss();
            });
        }
    }
}

File: Native.Android/app/src/main/java/com/pgrzmil/nativedroid/PerformanceTestActivity.java
package com.pgrzmil.nativedroid;

import android.app.ProgressDialog;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.app.Activity;
import com.pgrzmil.services.*;

public class PerformanceTestActivity extends Activity implements PerformanceTestListener {
    Stopwatch stopwatch;
    Button startButton;
    EditText digitsEntry;
    TextView timeLabel;
    TextView resultView;
    ProgressDialog progressDialog;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_performance);
        setTitle(R.string.menuPerformance);

        digitsEntry = (EditText) findViewById(R.id.digitsEntry);
        timeLabel = (TextView) findViewById(R.id.timeLabel);
        resultView = (TextView) findViewById(R.id.resultView);
        startButton = (Button) findViewById(R.id.startButton);

        stopwatch = new Stopwatch();
        PerformanceTestService.getInstance().addListener(this);
    }

    public void startCalculation(View view) {
        progressDialog = ProgressDialog.show(this, "Przetwarzanie...", "");
        stopwatch.start();

        final int digits = Integer.parseInt(digitsEntry.getText().toString());
        new Thread(new Runnable() {
            public void run() {
                PerformanceTestService.getInstance().calculatePi(digits);
            }
        }).start();
    }

    @Override
    public void piCalculationCompleted(final String result) {
        stopwatch.stop();
        runOnUiThread(new Runnable() {
            public void run() {
                resultView.setText(result);
                timeLabel.setText(stopwatch.getDurationInSeconds());
                progressDialog.dismiss();
            }
        });
    }
}
